lib/loader/loader.c

```diff
@@ -0,0 +1,118 @@
+#define DLL_OPEN(filename) LoadLibrary(filename)
+#define DLL_CLOSE(handle) FreeLibrary(handle)
+#define DLL_GET_CREATE_FN(handle, name) (evmc_create_fn) GetProcAddress(handle, name)
+#define HAVE_STRCPY_S 1
+#else
+#include <dlfcn.h>
+#define DLL_HANDLE void*
+#define DLL_OPEN(filename) dlopen(filename, RTLD_LAZY)
+#define DLL_CLOSE(handle) dlclose(handle)
+#define DLL_GET_CREATE_FN(handle, name) (evmc_create_fn)(uintptr_t) dlsym(handle, name)
+#define HAVE_STRCPY_S 0
+#endif
+
+#define PATH_MAX_LENGHT 4096
+
+#if !HAVE_STRCPY_S
+static void strcpy_s(char* dest, size_t destsz, const char* src)
+{
+    size_t len = strlen(src);
+    if (len > destsz - 1)
+        len = destsz - 1;
+    memcpy(dest, src, len);
+    dest[len] = 0;
+}
+#endif
+
+
+evmc_create_fn evmc_load(const char* filename, enum evmc_loader_error_code* error_code)
+{
+    enum evmc_loader_error_code ec = EVMC_ERRC_SUCCESS;
+    evmc_create_fn create_fn = NULL;
+
+    if (!filename)
+    {
+        ec = EVMC_ERRC_INVALID_ARGUMENT;
+        goto exit;
+    }
+
+    const size_t length = strlen(filename);
+    if (length == 0 || length > PATH_MAX_LENGHT)
+    {
+        ec = EVMC_ERRC_INVALID_ARGUMENT;
+        goto exit;
+    }
+
+    DLL_HANDLE handle = DLL_OPEN(filename);
+    if (!handle)
+    {
+        ec = EVMC_ERRC_CANNOT_OPEN;
+        goto exit;
+    }
+
+    // Create name buffer with the prefix.
+    const char prefix[] = "evmc_create_";
+    const size_t prefix_length = strlen(prefix);
+    char name[sizeof(prefix) + PATH_MAX_LENGHT];
+    strcpy_s(name, sizeof(name), prefix);
+
+    // Find filename in the path.
+    const char* sep_pos = strrchr(filename, '/');
+    const char* name_pos = sep_pos ? sep_pos + 1 : filename;
+
+    // Skip "lib" prefix if present.
+    const char lib_prefix[] = "lib";
+    const size_t lib_prefix_length = strlen(lib_prefix);
+    if (strncmp(name_pos, lib_prefix, lib_prefix_length) == 0)
+        name_pos += lib_prefix_length;
+
+    strcpy_s(name + prefix_length, PATH_MAX_LENGHT, name_pos);
+
+    // Trim the file extension.
+    char* ext_pos = strrchr(name, '.');
+    if (ext_pos)
+        *ext_pos = 0;
+
+    // Replace all "-" with "_".
+    char* dash_pos = name;
+    while ((dash_pos = strchr(dash_pos, '-')) != NULL)
+        *dash_pos++ = '_';
+
+    // Search for the "full name" based function name.
+    create_fn = DLL_GET_CREATE_FN(handle, name);
+    if (!create_fn)
+    {
+        // Try the "short name" based function name.
+        const char* short_name_pos = strrchr(name, '_');
+        if (short_name_pos)
+        {
+            short_name_pos += 1;
+            memmove(name + prefix_length, short_name_pos, strlen(short_name_pos) + 1);
+            create_fn = DLL_GET_CREATE_FN(handle, name);
+        }
+    }
+
+    if (!create_fn)
+    {
+        DLL_CLOSE(handle);
+        ec = EVMC_ERRC_SYMBOL_NOT_FOUND;
+    }
+
+exit:
+    if (error_code)
+        *error_code = ec;
+    return create_fn;
```

Comment:
So `handle` stays open forever after sucessful `evmc_load`, is this ok?

Reply:
I don't think you can close it if there's a pointer into it?

Reply:
On Windows there is `FreeLibrary` https://msdn.microsoft.com/en-us/library/windows/desktop/ms683152(v=vs.85).aspx

Reply:
The `dlclose()` will unload the library from the memory so you cannot use functions from it (similarly on Windows). We are doing so only when we could not find the create function because then the library is not needed any more. In case we use the library, it is never unloaded.